doc: Add information on Partition Manager and external devices (#9423)
nrfconnect/sdk-nrf: scripts/partition_manager/partition_manager.rst

.. _pm_build_system:

> For a partition to be accessible at run-time, a flash device it is placed on needs to have its driver enabled at compile time. Partition manager will not includ, into partition map, partitions that are placed on a device without a driver in the system.

Reviewer: ```suggestion
For a partition to be accessible at run-time, a flash device it is placed on needs to have its driver enabled at compile time.
Partition manager will ignore partitions that are located in a region without a driver associated with it.
```
That is the case, right? Given that a region has a device associated with it.

> To let the partition manager know which Kconfig ensures existence of a driver, the ``DEFAULT_DRIVER_KCONFIG`` option is used.

Reviewer: Is this an option? If its kconfig it should be prefixed with `CONFIG_` if its not kconfig, I think we should avoid calling it at "option" as that is typically used for kconfig options, and instead call it what it is (for example "cmake variable" if thats the case)

Author: This is param to Partition Manager script.

Author: Used https://github.com/nrfconnect/sdk-nrf/pull/9423#discussion_r1033976314 here

> For a partition to be accessible at run-time, a flash device it is placed on needs to have its driver enabled at compile time. Partition manager will not includ, into partition map, partitions that are placed on a device without a driver in the system.
> To let the partition manager know which Kconfig ensures existence of a driver, the ``DEFAULT_DRIVER_KCONFIG`` option is used.
> Within :file:`partition_manager.cmake` the options is assigned :kconfig:option:`CONFIG_SOC_FLASH_NRF` for partitions on internal flash, which means that all partitions placed on internal flash area always available at run-time.

Reviewer 2: ```suggestion
A partition can be accessible during runtime only if the flash device that contains the partition has its own driver enabled during compile time. 
Partition manager does not include, into the partition map, partitions that are placed on a device without a driver in the system.
To let the partition manager know which Kconfig options ensure the existence of a driver, use ``DEFAULT_DRIVER_KCONFIG``.
```

Author: Applied

For external regions the :file:`partition_manager.cmake` assigns the ``DEFAULT_DRIVER_KCONFIG`` with :kconfig:option:`CONFIG_PM_EXTERNAL_FLASH_HAS_DRIVER`; the option needs to be set by external driver or Kconfig.

> This option is automatically set when :kconfig:option:`CONFIG_NRF_QSPI_NOR` or :kconfig:option:`CONFIG_SPI_NOR` is enabled, and can be override with setting :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` in application configuration.

Reviewer 2: ```suggestion
This option is automatically set when :kconfig:option:`CONFIG_NRF_QSPI_NOR` or :kconfig:option:`CONFIG_SPI_NOR` are enabled, and can be overridden by setting :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` in the application configuration.
```

.. _pm_build_system:

For a partition to be accessible at run-time, a flash device it is placed on needs to have its driver enabled at compile time.
Partition manager will ignore partitions that are located in a region without a driver associated with it.

> To let the partition manager know which Kconfig ensures existence of a driver, the ``DEFAULT_DRIVER_KCONFIG`` option is used.

Reviewer 2: ```
A partition can be accessible at runtime only if the flash device where it resides has its driver enabled at compile time.
Partition manager ignores partitions that are located in a region without its driver enabled.
To let partition manager know which Kconfig option ensures the existence of the driver, the option ``DEFAULT_DRIVER_KCONFIG`` is used.
```

Author: Taken

> Within :file:`partition_manager.cmake` the options is assigned :kconfig:option:`CONFIG_SOC_FLASH_NRF` for partitions on internal flash, which means that all partitions placed on internal flash area always available at run-time.

Reviewer 3: ```suggestion
Within the :file:`partition_manager.cmake` file the option is assigned the :kconfig:option:`CONFIG_SOC_FLASH_NRF` option for partitions in internal flash, which means that all partitions placed in internal flash are always available at run-time.
```

Reviewer 3: There are other locations in your partition_manager.rst that also need to add " file" after any :file: reference, and " option" after any :kconfig:option: reference, as I did in the suggestion.

> Within the :file:`partition_manager.cmake` file the option is assigned the :kconfig:option:`CONFIG_SOC_FLASH_NRF` option for partitions in internal flash, which means that all partitions placed in internal flash are always available at run-time.
> For external regions the :file:`partition_manager.cmake` assigns the ``DEFAULT_DRIVER_KCONFIG`` with :kconfig:option:`CONFIG_PM_EXTERNAL_FLASH_HAS_DRIVER`; the option needs to be set by external driver or Kconfig.
> This option is automatically set when :kconfig:option:`CONFIG_NRF_QSPI_NOR` or :kconfig:option:`CONFIG_SPI_NOR` are enabled, and can be overridden by setting :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` in the application configuration.

Reviewer 2: ```suggestion
For partitions in the internal flash memory, ``DEFAULT_DRIVER_KCONFIG`` within :file:`partition_manager.cmake` is set automatically to :kconfig:option:`CONFIG_SOC_FLASH_NRF`, so that all the partitions placed in the internal flash are always available at runtime.
For external regions, ``DEFAULT_DRIVER_KCONFIG`` within :file:`partition_manager.cmake` must be set to :kconfig:option:`CONFIG_PM_EXTERNAL_FLASH_HAS_DRIVER`.
Out-of-tree drivers can select this value to attest that they provide support for the external flash.
This is an hidden option and can be selected only by an external driver or a Kconfig option.

This option is automatically set when :kconfig:option:`CONFIG_NRF_QSPI_NOR` or :kconfig:option:`CONFIG_SPI_NOR` are enabled.
If the application provides the driver in an unusual way, this option can be overridden by setting :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` in the application configuration.
```

> As stated before, the partition manager will only include into compilation partitions that it is able to match with a driver, as pointed by the ``DEFAULT_DRIVER_KCONFIG``.
> Partition manager does not know whether partitions are used at run-time or not, it only knows whether they have driver compiled in or not.
> 
> This has following effects:

Reviewer 2: ```suggestion
As partition manager does not know if partitions are used at runtime, consider the following:
```

>   * Enabling Kconfig that affect the ``DEFAULT_DRIVER_KCONFIG`` will add partition map entry for the partition depending on it, whether it is used in run-time or not.
> 
>   * Not enabling the Kconfig that affect the ``DEFAULT_DRIVER_KCONFIG`` will not add partition map entry for partition depending on it, whether it is used in run-time or not.
> 
>   * Enabling Kconfig options that affects the ``DEFAULT_DRIVER_KCONFIG`` may cause linker errors when the option has no effect on including driver into compilation. In such case partition manager adds partition map entry that has a pointer to flash device it is supposed to be placed on, but due to misconfiguration a driver is actually not compiled in.
>     Such situation may also be caused by setting the :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` that will just assume that driver is provided by application.

Reviewer 2: ```suggestion
  * Enabling Kconfig options that affect ``DEFAULT_DRIVER_KCONFIG`` will add a partition map entry for the partition depending on it, whether it is used at runtime or not.
  * Not enabling the Kconfig options that affect ``DEFAULT_DRIVER_KCONFIG`` will not add partition map entry for partition depending on it, whether it is used at runtime or not.
  * Enabling Kconfig options that affects ``DEFAULT_DRIVER_KCONFIG`` can cause linker errors when the option has no effect on including a driver into compilation.
    In this case, partition manager adds a partition map entry that has a pointer to the flash device it is supposed to be placed on, but due to misconfiguration the driver is actually not compiled in.
    This situation can also be caused by setting the :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK`, as partition manager will just assume that the driver is provided by the application.
```

.. note::

>    When using an app configured with an MCUboot child image, both images use the same partition manager configuration, which means that the app and MCUboot will have exactly the same partition maps. Whether flash partitions will be accessible for them at run-time depends on application and MCUboot configurations and the values they give to the ``DEFAULT_DRIVER_KCONFIG`` option of the partition manager region specification.

Reviewer 2: ```suggestion
   When using an application configured with an MCUboot child image, both images use the same partition manager configuration, which means that the app and MCUboot have exactly the same partition maps.
   The accessibility at runtime of flash partitions depend on the configurations of both the application and MCUboot and the values they give to the ``DEFAULT_DRIVER_KCONFIG`` option of the partition manager region specification.
```

>    When using app configuration with MCUboot child image, both images use the same partition manager configuration, which means that app and MCUboot will have exectly the same partition maps generated, but whether flash partitions will be accessible for them, at run-time, depends on application and MCUboot configurations and values they give to the Kconfig options, that are given to the ``DEFAULT_DRIVER_KCONFIG`` option of pratition manager region specification.

Reviewer 3: ```suggestion
   When using an app configured with an MCUboot child image, both images use the same partition manager configuration, which means that the app and MCUboot will have exactly the same partition maps. Whether flash partitions will be accessible for them at run-time depends on application and MCUboot configurations and the values they give to the ``DEFAULT_DRIVER_KCONFIG`` option of the partition manager region specification.
```
